### ISSUE TITLE
Importing from other dart packages is broken

### ISSUE THREAD

**Issue by yury-yufimov** [AdalbertoLacruz/less_dart#24]
less_dart version: 2.3.2

You can't use imports from other dart-packages:  
after css generation is completed, builder will try to mark dependency via
```
// dependents - force AssetGraph inclusion, we are called on change
transformer.imports.forEach((String path) async {
  await buildStep.canRead(AssetId(package, path));
});
```
but `path` is an absolute path, and `package` is not the packages you're importing

example:
https://github.com/yury-yufimov/less_dart_issue_example/blob/master/main_pkg/README.md

**Comment by AdalbertoLacruz** [comment]
OK.

Packages came to this point in absolute. The immediate solution is to ignore them.  Packages must not change continuously, so this must not be a problem. Any case, editing the file that is importing the package must resolve to a full build.

**Comment by AdalbertoLacruz** [comment]
Version 2.3.3 is in pub. 

It supports yet changes inside a package.

**Comment by yury-yufimov** [comment]
@AdalbertoLacruz , thank you for quick response!  
However, I still can't build my example project with 2.3.3 version (and have troubles with debugging via webdev). 

Second unclear moment is relative paths for imports. Does less_dart allows it? What about relative imports not in root package?
(As you know, relative imports within the package are more practical and allow navigation via IDE)

Here's some small example of what I'm talking about:
https://github.com/yury-yufimov/less_dart_issue_example/tree/complicated-case

As to my expectations:
our company uses forked version of less_dart's builder, which allows all of the above. But it's based on old version of less_dart and requires some extra-configuration.
Now we are looking into refactoring of our build system to use webdev. That complicates usage of forked library versions, so we are looking into migration on unchanged less_dart.
Long story short: 
1) there's no need for urgent fix
2) I'm convinced, that any large project sooner or later will require such powerful import-system.
Minimum requirements, in my opinion:
support of https://github.com/yury-yufimov/less_dart_issue_example/tree/master  
Maximal:
support of https://github.com/yury-yufimov/less_dart_issue_example/tree/complicated-case

**Comment by AdalbertoLacruz** [comment]
It works.

You need a build.yaml file:

```
targets:
  $default:
    builders:
      less_dart:
        options:
          entry_points: ['web/app_component.less']
          include_path: 'lib/structure/common'

```
The main .less files to convert must be under web directory, to be included in build.

Change in app_component.less:

```
/* @import (reference) "structure/component/component.less"; */
@import (reference) "packages/main_pkg/structure/component/component.less";
```

use> webdev build
And you get build/app_component.css. The angular matters are your business.

I Hope this resolves your issue.

**Comment by yury-yufimov** [comment]
@AdalbertoLacruz , thanks.  
I confirm, that now task of "migrating our codebase to less_dart 2.3.3" is doable.  
I think the issue could be completed.

Nevertheless, I believe that absolute (package) imports are not easy to use: IDE can't perform navigation, code moving/refactoring is more complicated.
Secondly, include_path approach for relative paths has problems too: IDE can't perform navigation either, and someday you'll have 2 files with same name in different include_paths...

What do you think about pull-request, where I'll try to migrate our approach to imports? It supports imports, relative to current less-file, with no need of extra include_path.
Yes, this is more convenient for angular applications where you have no project-wide style files, only small ones for components.
And yes, I think this should be major update - handling of relative imports will be changed.

Or that conflicts with you vision of less_dart package purpose?

**Comment by AdalbertoLacruz** [comment]
Ending the before point, you could organize yours components as
```
web/components/button/button.less
...
web/components/dialog/dialog.less
```

and define
`entry_points: ['web/app_component.less', 'web/components/*/*.less']`

----------------------------

About paths, always are relative. 
When you import from a package, you define the file starting from package, lib directory.
Inside the package the imports are also relative.
Your example has relative imports and it works.

I don't understand which is the problem.

**Comment by yury-yufimov** [comment]
@AdalbertoLacruz , I think, slowly I'm starting to understand current system for relative imports. Can you check, that I'm right?  

1) By default, all *.less files are entrypoints
2) Any less files can have both: package(absolute) imports and relative imports
3) (!!!) relative imports in entrypoint are searched from project root, not from entrypoint location
4) Other less files, (transitively-)imported from entrypoint can have relative imports - that imports are searched from file location.

Example to demonstrate this structure:
https://github.com/yury-yufimov/less_dart_issue_example/blob/relative-paths-in-entrypoint/main_pkg/lib/app_component.less
(project is working now!)

If I'm right, I'll try to explain, why this difference in relative-imports handling for entrypoints and other files is hard to deal with:
1) Support of entrypoint list in big projects will require either too strict file structure (e.g. all components in web/components/*/ or lib/components/*/) or a lot of manual work, which will tend to errors or garbage in config files
2) Some less files can simultaneously be entrypoint in one project and (transitive-)import in another. Such files can't have relative imports at all.
3) IDE can't perform navigation to relative imports in entrypoints: them are searched from project root by less_dart and from file location by IDE

**Comment by AdalbertoLacruz** [comment]
The way less_dart import files is the same as in the original/official javascript version. Also, they have an equivalent for package imports, from npm packages.

Yours points:

1. In the builder (webdev build) we need to define a build.yaml file. In that file you define the entry points, as in the example:

`entry_points: ['web/app_component.less', 'web/components/*/*.less']`

That define only web/app_component.less, and all the .less files below web/components.

So, NOT all .less files are entry points, only the files you define, one by one or by wildcards. If you want all .less files:

`entry_ponts: ['*.less']`

That process all .less files that the builder receive from the dart builder system.

2. Any less file could import files, relative to the path for the same file that is importing, and from packages. In the tests, you have many examples. See

`https://github.com/AdalbertoLacruz/less_dart/tree/master/test/less/import`

3/4. Relative imports are from the file that is importing.

About file structure:
The dart system (webdev build) copies all in  \web directory to \build directory.
In the process of building it transform some files (so .less to .css if it is defined). If you define an entry point in the lib directory, LESS transform it, but the builder don't let write it to the lib directory. The builder only writes to the build directory, but delivers the builded file to the next builder. 

The build.yaml file must define the less builder, and then, the angular builder. That lib/components/*/*.less  transformed to lib/components/*/*.css possibly could be the input for the angular builder. That is your fun.

**Comment by yury-yufimov** [comment]
1. I've just wanted to say, that if you declare no entry-points, it will be the same, as to declare
`entry_ponts: ['*.less']`
(due to https://github.com/AdalbertoLacruz/less_dart/blob/master/lib/src_builder/entry_points.dart#L7 )

2. The way less is executed in tests is not the same, as it executed for entry_point.
In tests you pass file name to transform:
![Screen Shot 2019-10-31 at 12 54 57](https://user-images.githubusercontent.com/33983822/67941777-76c2b600-fbe7-11e9-9cea-8ce943edbbc5.png)

But when less is executed for entrypoint, no path/filename is given to transform (`-` is used instead)
![Screen Shot 2019-10-31 at 14 10 45](https://user-images.githubusercontent.com/33983822/67942171-4f201d80-fbe8-11e9-9c46-6d3e22a00b51.png)

> 3/4. Relative imports are from the file that is `importing.

No, that's not true for entrypoints. I've modified an example a bit, to show it more precisely:
https://github.com/yury-yufimov/less_dart_issue_example/blob/relative-paths-in-entrypoint/main_pkg/web/components/component_b/component_b.less

I think this is just a bug, and the only bug that stops us from migrating - handling of relative imports in entrypoints.

I do understand that build process place generated files not in lib folder, but in .dart_tool/build/generated , but that's totally fine to use with angular: them are visible to it's builder and converted into *.css.shim.dart 
Btw, you can configure builder to place results near source-files, via
`build_to: source|cache` option

**Comment by AdalbertoLacruz** [comment]
OK. You are right. Thanks for the debugging.
Builder transforms the .less file by stdin, losing the filepath information.
Hope the new version 2.3.4 resolves this.

**Comment by yury-yufimov** [comment]
@AdalbertoLacruz , that's great!
Just for double-checking: don't you think it should be a major update? If any one uses previous behaviour (have imports in entry-point, relative to project-root), it will be broken.
That won't affect our project, I'm just wondering.

**Comment by AdalbertoLacruz** [comment]
It supports previous behavior. You could use also imports relative to project-root.

And, if you define

`include_path: 'lib'`

relative to \lib.

**Comment by yury-yufimov** [comment]
@AdalbertoLacruz , sorry for delay - I haven't notice new version.  

Unfortunately, I still can't report the issue is solved (
Current non-working case is relative imports not in root, but in dependent package

I've created example for that:
https://github.com/yury-yufimov/less_dart_issue_example/tree/relative-imports-in-dependency

You can successfully run `webdev serve` in components_pkg, but not in main_pkg.

As I can see, relative imports are always calculated from root package now.

Or I've just misconfigured build.yaml?

**Comment by yury-yufimov** [comment]
@AdalbertoLacruz , here is my vision of needed fix

**Comment by AdalbertoLacruz** [comment]
2.3.5 in pub.

Build in other packages was not supported. Now I think it is.

Thanks for your version, but this one was yet tested. And supports build outside packages directories as 'web\...'.

By the way. I don't like generate useless files. You could move generic files to their own directory (lib/common/) and use the exclusion format:

`entry_points: ['*.less', '!/common/*.less']`

Obviously the imports must reference the common directory. So for button.less:

`@import (reference) "../common/palette.less";`

Other option is to define an include_path as root for the includes:
 

        options:
          entry_points: ['*.less', '!/common/*.less']
          include_path: 'packages/components_pkg/common'

and in button.less

`@import (reference) "palette.less";`

Hope, this resolves the issue.

**Comment by yury-yufimov** [comment]
@AdalbertoLacruz , I confirm that version 2.3.5 resolves the issue and is suitable for our usage with angular.

Thanks a lot! That was very interesting.
